sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/batch/ServerBatchRequest.java

```diff
@@ -0,0 +1,115 @@
+import com.azure.cosmos.implementation.Utils;
+import com.azure.cosmos.implementation.apachecommons.collections.list.UnmodifiableList;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+
+import java.util.List;
+
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkNotNull;
+import static com.azure.cosmos.implementation.guava25.base.Preconditions.checkState;
+
+/**
+ * This class represents a server batch request.
+ */
+public abstract class ServerBatchRequest {
+
+    private final int maxBodyLength;
+    private final int maxOperationCount;
+
+    private String requestBody;
+    private List<ItemBatchOperation<?>> operations;
+    private boolean isAtomicBatch = false;
+    private boolean shouldContinueOnError = false;
+
+    /**
+     * Initializes a new {@link ServerBatchRequest request} instance.
+     *
+     * @param maxBodyLength Maximum length allowed for the request body.
+     * @param maxOperationCount Maximum number of operations allowed in the request.
+     */
+    ServerBatchRequest(int maxBodyLength, int maxOperationCount) {
+        this.maxBodyLength = maxBodyLength;
+        this.maxOperationCount = maxOperationCount;
+    }
+
+    /**
+     * Adds as many operations as possible from the given list of operations.
+     * TODO(rakkuma): Similarly for hybrid row, request needs to be parsed to create a request body in any form.
+     * Issue: https://github.com/Azure/azure-sdk-for-java/issues/15856
+     *
+     * Operations are added in order while ensuring the request body never exceeds {@link #maxBodyLength}.
+     *
+     * @param operations operations to be added; read-only.
+     *
+     * @return Any pending operations that were not included in the request.
+     */
+    final List<ItemBatchOperation<?>> createBodyOfBatchRequest(final List<ItemBatchOperation<?>> operations) {
+
+        checkNotNull(operations, "expected non-null operations");
+
+        int totalSerializedLength = 0;
+        int totalOperationCount = 0;
+
+        final ArrayNode arrayNode =  Utils.getSimpleObjectMapper().createArrayNode();
+
+        for(ItemBatchOperation<?> operation : operations) {
+
+            final JsonSerializable operationJsonSerializable = ItemBatchOperation.writeOperation(operation);
+            final int operationSerializedLength = operationJsonSerializable.toString().length();
+
+            if (totalOperationCount != 0 &&
+                (totalSerializedLength + operationSerializedLength > this.maxBodyLength || totalOperationCount + 1 > this.maxOperationCount)) {
+                // Apply the limit only if at least there is one operation in selected operations.
+                break;
+            }
+
+            totalSerializedLength += operationSerializedLength;
+            totalOperationCount++;
+
+            arrayNode.add(operationJsonSerializable.getPropertyBag());
+        }
+
+        this.requestBody = arrayNode.toString();
```

Comment:
(optimization for later, out of scope of this PR, please track this).
here you are translating the jsonArrayNode to String. later on the encoding over wire you are translating the string to byte array/bytebuf.

The more optimized way is to translate directly to byte array.

Reply:
Oh. Ok. Will make this change. This is a small one, so will make it here only.

Reply:
I am saving ByteBuf now instead of byte[]. returning byte[] gives following error: 
'com.azure.cosmos.implementation.batch.ServerBatchRequest.getRequestBodyAsByteArray() may expose internal representation by returning ServerBatchRequest.requestBody [com.azure.cosmos.implementation.batch.ServerBatchRequest] At ServerBatchRequest.java:[line 90] EI_EXPOSE_REP'

We are not modifying ByteBuf later right?